clouddriver-artifacts/src/main/java/com/netflix/spinnaker/clouddriver/artifacts/gitRepo/GitJobExecutor.java

```diff
@@ -0,0 +1,294 @@
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.FileUtils;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.util.StringUtils;
+
+@Slf4j
+public class GitJobExecutor {
+
+  private static final String SSH_KEY_PWD_ENV_VAR = "SSH_KEY_PWD";
+  private static Path genericAskPassBinary;
+
+  @Getter private final GitRepoArtifactAccount account;
+  private final JobExecutor jobExecutor;
+  private final String gitExecutable;
+  private final AuthType authType;
+  private final Path askPassBinary;
+
+  private enum AuthType {
+    USER_PASS,
+    TOKEN,
+    SSH,
+    NONE
+  }
+
+  public GitJobExecutor(
+      GitRepoArtifactAccount account, JobExecutor jobExecutor, String gitExecutable)
+      throws IOException {
+    this.account = account;
+    this.jobExecutor = jobExecutor;
+    this.gitExecutable = gitExecutable;
+    if (!StringUtils.isEmpty(account.getUsername())
+        && !StringUtils.isEmpty(account.getPassword())) {
+      authType = AuthType.USER_PASS;
+    } else if (!StringUtils.isEmpty(account.getToken())) {
+      authType = AuthType.TOKEN;
+    } else if (!StringUtils.isEmpty(account.getSshPrivateKeyFilePath())) {
+      authType = AuthType.SSH;
+    } else {
+      authType = AuthType.NONE;
+    }
+    askPassBinary = initAskPass();
+  }
+
+  public void clone(String repoUrl, String branch, Path destination) throws IOException {
+    if (!isValidReference(repoUrl)) {
+      throw new IllegalArgumentException(
+          "Git reference \""
+              + repoUrl
+              + "\" is invalid for credentials with auth type "
+              + authType);
+    }
+    FileUtils.deleteDirectory(destination.toFile());
+    FileUtils.forceMkdir(destination.toFile());
+    log.info("Cloning git/repo {} into {}", repoUrl, destination.toString());
+
+    String cloneCommand =
+        gitExecutable + " clone --branch " + branch + " --depth 1 " + repoUrlWithAuth(repoUrl);
+
+    List<String> command = cmdToList(cloneCommand);
+    log.debug("Executing command: \"{}\"", String.join(" ", command));
+
+    JobResult<String> result =
+        jobExecutor.runJob(
+            new JobRequest(command, addEnvVars(System.getenv()), destination.toFile()));
+
+    if (result.getResult() != JobResult.Result.SUCCESS) {
+      throw new IOException(
+          "Failed to clone repository "
+              + repoUrl
+              + " into "
+              + destination
+              + ". Error: "
+              + result.getError()
+              + " Output: "
+              + result.getOutput());
+    }
+  }
+
+  public void archive(Path localClone, String branch, String subDir, Path outputFile)
+      throws IOException {
+
+    List<String> command = new ArrayList<>();
+    command.add(gitExecutable);
+    command.add("archive");
+    command.add("--format");
+    command.add("tgz");
+    command.add("--output");
+    command.add(outputFile.toString());
+    command.add(branch);
```

Comment:
Small style nit - this could be simplified using `Arrays.asList`

```suggestion
List<String> command = Arrays.asList(
  gitExecutable , "archive", "--format", "tgz", "--output", outputFile.toString(), branch);
```